File: AiSD5_C2/exercise_4/Python/main.py
import math
import time
from datetime import datetime, timedelta

INSTALLED_POWER = 3.0  # 3 kW installed power
SIMULATION_SECONDS = 100  # Simulation time in seconds (instead of 365 days)
REACTIVE_POWER_RATIO = 0.1  # Reactive power is 10% of active power

class PowerData:
    def __init__(self):
        self.active_power_sum1 = 0.0
        self.active_power_sum2 = 0.0
        self.reactive_power_sum1 = 0.0
        self.reactive_power_sum2 = 0.0
        self.max_active_power = 0.0
        self.last_max_power_date = datetime.now()
        self.last_month_max_power = 0.0
        self.start_date = datetime.now()

def is_in_tariff_i(hour):
    return (6 <= hour < 13) or (15 <= hour < 22)

def simulate_active_power(time):
    return (math.sin(time * math.pi / 180) + 1) * INSTALLED_POWER / 2

def process_power_readings(data, active_power, reactive_power, current_time, current_hour):
    if is_in_tariff_i(current_hour):
        data.active_power_sum1 += active_power
        data.reactive_power_sum1 += reactive_power
    else:
        data.active_power_sum2 += active_power
        data.reactive_power_sum2 += reactive_power

    current_month = current_time.month
    last_max_month = data.last_max_power_date.month

    if current_month != last_max_month or data.last_month_max_power == 0:
        data.last_month_max_power = data.max_active_power
        data.max_active_power = active_power
        data.last_max_power_date = current_time
    elif active_power > data.max_active_power:
        data.max_active_power = active_power
        data.last_max_power_date = current_time

def display_results(data):
    print('=== Raport zużycia energii ===')
    print(f'Moc Czynna (Taryfa I): {data.active_power_sum1:.2f} kWh')
    print(f'Moc Bierna (Taryfa I): {data.reactive_power_sum1:.2f} kVArh')
    print(f'Moc Czynna (Taryfa II): {data.active_power_sum2:.2f} kWh')
    print(f'Moc Bierna (Taryfa II): {data.reactive_power_sum2:.2f} kVArh')
    print(f'Moc maksymalna w ostatnim miesiącu: {data.last_month_max_power:.2f} kW')
    print(f'Moc maksymalna w obecnym miesiącu: {data.max_active_power:.2f} kW')
    print('==============================')

def control_program(data):
    for elapsed_seconds in range(SIMULATION_SECONDS + 1):
        current_time = data.start_date + timedelta(seconds=elapsed_seconds)
        current_hour = current_time.hour

        active_power = simulate_active_power(elapsed_seconds)
        reactive_power = active_power * REACTIVE_POWER_RATIO

        process_power_readings(data, active_power, reactive_power, current_time, current_hour)
        display_results(data)

        time.sleep(0.1)

if __name__ == "__main__":
    power_data = PowerData()
    control_program(power_data)
    input('Wcisnij Enter aby wyjść z programu...')

File: AiSD5_C2/exercise_4/Pascal/main.pas
program PowerConsumption;
uses
  SysUtils, Math, DateUtils, Crt;

const
  INSTALLED_POWER = 3.0; // 3 kW installed power
  SIMULATION_SECONDS = 100; // Simulation time in seconds (instead of 365 days)
  REACTIVE_POWER_RATIO = 0.1; // Reactive power is 10% of active power

type
  TPowerData = record
    ActivePowerSum1: Double;    // Sum of active power in tariff I
    ActivePowerSum2: Double;    // Sum of active power in tariff II
    ReactivePowerSum1: Double;  // Sum of reactive power in tariff I
    ReactivePowerSum2: Double;  // Sum of reactive power in tariff II
    MaxActivePower: Double;     // Maximum active power in current month
    LastMaxPowerDate: TDateTime;// Date of last maximum power recording
    LastMonthMaxPower: Double;  // Maximum active power from previous month
    StartDate: TDateTime;       // Counter start date
  end;

// Function to check if current time is in Tariff I (6:00-13:00 or 15:00-22:00)
function IsInTariffI(Hour: Integer): Boolean;
begin
  IsInTariffI := ((Hour >= 6) and (Hour < 13)) or ((Hour >= 15) and (Hour < 22));
end;

// Function to simulate instantaneous power consumption
function SimulateActivePower(Time: Integer): Double;
begin
  // Simulate power consumption using sine function (between 0 and INSTALLED_POWER)
  SimulateActivePower := (Sin(Time * PI / 180) + 1) * INSTALLED_POWER / 2;
end;

// Procedure to process power readings
procedure ProcessPowerReadings(var Data: TPowerData; ActivePower, ReactivePower: Double; 
                             CurrentTime: TDateTime; CurrentHour: Integer);
var
  CurrentMonth, LastMaxMonth: Integer;
begin
  // Update power sums based on tariff
  if IsInTariffI(CurrentHour) then
  begin
    Data.ActivePowerSum1 := Data.ActivePowerSum1 + ActivePower;
    Data.ReactivePowerSum1 := Data.ReactivePowerSum1 + ReactivePower;
  end
  else
  begin
    Data.ActivePowerSum2 := Data.ActivePowerSum2 + ActivePower;
    Data.ReactivePowerSum2 := Data.ReactivePowerSum2 + ReactivePower;
  end;

  CurrentMonth := MonthOf(CurrentTime);
  LastMaxMonth := MonthOf(Data.LastMaxPowerDate);

  // Update maximum power values
  if (CurrentMonth <> LastMaxMonth) or (Data.LastMonthMaxPower = 0) then
  begin
    Data.LastMonthMaxPower := Data.MaxActivePower;
    Data.MaxActivePower := ActivePower;
    Data.LastMaxPowerDate := CurrentTime;
  end
  else if (ActivePower > Data.MaxActivePower) then
  begin
    Data.MaxActivePower := ActivePower;
    Data.LastMaxPowerDate := CurrentTime;
  end;
end;

// Procedure to display results
procedure DisplayResults(const Data: TPowerData);
begin
  ClrScr;
  WriteLn('=== Raport zużycia energii ===');
  WriteLn(Format('Moc Czynna (Taryfa I): %.2f kWh', [Data.ActivePowerSum1]));
  WriteLn(Format('Moc Bierna (Taryfa I): %.2f kVArh', [Data.ReactivePowerSum1]));
  WriteLn(Format('Moc Czynna (Taryfa II): %.2f kWh', [Data.ActivePowerSum2]));
  WriteLn(Format('Moc Bierna (Taryfa II): %.2f kVArh', [Data.ReactivePowerSum2]));
  WriteLn(Format('Moc maksymalna w ostatnim miesiącu: %.2f kW', [Data.LastMonthMaxPower]));
  WriteLn(Format('Moc maksymalna w obecnym miesiącu: %.2f kW', [Data.MaxActivePower]));
  WriteLn('==============================');
end;

// Main control procedure
procedure ControlProgram(var Data: TPowerData);
var
  CurrentTime: TDateTime;
  ElapsedSeconds: Integer;
  ActivePower, ReactivePower: Double;
  CurrentHour: Integer;
begin
  // Initialize data
  Data.ActivePowerSum1 := 0;
  Data.ActivePowerSum2 := 0;
  Data.ReactivePowerSum1 := 0;
  Data.ReactivePowerSum2 := 0;
  Data.MaxActivePower := 0;
  Data.LastMonthMaxPower := 0;
  Data.StartDate := Now;
  Data.LastMaxPowerDate := Data.StartDate;

  // Main simulation loop
  for ElapsedSeconds := 0 to SIMULATION_SECONDS do
  begin
    CurrentTime := IncSecond(Data.StartDate, ElapsedSeconds);
    CurrentHour := HourOf(CurrentTime);
    
    // Get power readings
    ActivePower := SimulateActivePower(ElapsedSeconds);
    ReactivePower := ActivePower * REACTIVE_POWER_RATIO;

    // Process readings
    ProcessPowerReadings(Data, ActivePower, ReactivePower, CurrentTime, CurrentHour);
    
    // Display current state
    DisplayResults(Data);
    
    // Add small delay to simulate real-time processing
    Sleep(100);
  end;
end;

var
  PowerData: TPowerData;

begin
  ControlProgram(PowerData);
  ReadLn; // Wait for user input before closing
end.